lib/rubygems/compiler.rb

```diff
@@ -47,6 +47,8 @@ def compile
       gemspec.files.push file
     end
 
+    gemspec.files = gemspec.files.select {|f| File.exists?("#{target_dir}/#{f}") } if @options[:purge]
```

Comment:
Why not use `reject!` instead? `gemspec.files` is an Array. I also recommend use `File.exist?` instead of the plural/awkward `exists?`.

Also do this before adding the artifact to the gemspec, that way you're not stating for a file that already exists (the recent added artifacts).